draft-ietf-quic-http.md

```diff
@@ -682,36 +682,38 @@ keep connections open.
 ## Connection Shutdown
 
 Even when a connection is not idle, either endpoint can decide to stop using the
-connection and let the connection close gracefully.  Since clients drive request
-generation, clients perform a connection shutdown by not sending additional
-requests on the connection; responses and pushed responses associated to
-previous requests will continue to completion.  Servers perform the same
-function by communicating with clients.
-
-Servers initiate the shutdown of a connection by sending a GOAWAY frame
-({{frame-goaway}}).  The GOAWAY frame indicates that client-initiated requests
-on lower stream IDs were or might be processed in this connection, while
-requests on the indicated stream ID and greater were rejected. This enables
-client and server to agree on which requests were accepted prior to the
-connection shutdown.  This identifier MAY be zero if no requests were processed.
-Servers SHOULD NOT permit additional QUIC streams after sending a GOAWAY frame.
-
-Clients MUST NOT send new requests on the connection after receiving GOAWAY;
-a new connection MAY be established to send additional requests.
-
-Some requests might already be in transit. If the client has already sent
-requests on streams with a Stream ID greater than or equal to that indicated in
-the GOAWAY frame, those requests will not be processed and MAY be retried by the
-client on a different connection.  The client MAY cancel these requests.  It is
-RECOMMENDED that the server explicitly reject such requests (see
-{{request-cancellation}}) in order to clean up transport state for the affected
-streams.
-
-Requests on Stream IDs less than the Stream ID in the GOAWAY frame might have
-been processed; their status cannot be known until a response is received, the
-stream is reset individually, or the connection terminates.  Servers MAY reject
-individual requests on streams below the indicated ID if these requests were not
-processed.
+connection and initiate a graceful connection close.  Endpoints initiate the
+graceful shutdown of a connection by sending a GOAWAY frame ({{frame-goaway}}).
+The GOAWAY frame contains an identifier that indicates to the receiver the range
+of requests or pushes that were or might be processed in this connection.  The
+server sends a client-initiated bidirectional Stream ID; the client sends a Push
+ID.  Requests or pushes with the indicated identifier or greater are rejected by
+the sender of the GOAWAY.  This identifier MAY be zero if no requests or pushes
+were processed.
+
+The information in the GOAWAY enables a client and server to agree on which
+requests or pushes were accepted prior to the connection shutdown.  Endpoints
+SHOULD abruptly terminate any requests or pushes that have identifiers greater
+than or equal to the smallest identifier sent in a GOAWAY frame.
```

Comment:
I think we can remove this sentence, as we talk more specifically in the sentence starting from line 708 (i.e. "It is RECOMMENDED that the receiving endpoint explicitly reject such requests (see {{request-cancellation}}) or pushes (see {{frame-cancel-push}}) in order to clean up transport state for the affected streams."), and because the abstract is already given in the sentence starting from line 690 (i.e. "Requests or pushes with the indicated identifier or greater are rejected by the sender of the GOAWAY.").

Reply:
This is talking about the sending endpoint, though.  I've received requests A-G; I'm sending a GOAWAY indicating that I've done some non-zero amount of processing up through C.  I send a GOAWAY indicating that D and higher haven't been processed, then SHOULD send RESET_STREAMs for D, E, F, and G to keep the transport state consistent.

Reply:
> This is talking about the sending endpoint, though.

That's true, and I now notice that the sentence starting from line 708 is the one that's incorrect.

The paragraph starting from line 703 talks about what to do with messages that were already in transit. It is a generalization of an exiting paragraph, that states that the sender of GOAWAY SHOULD reset, the receiver of GOAWAY MAY. I assume that we want to retain those behaviors.

The second sentence of that paragraph that talks about the receiver is correct (i.e. "If the endpoint has already sent requests or pushes with an identifier greater than or equal to that received in a GOAWAY frame, those requests or pushes will not be processed"). Next sentence that starts from 708 should then talk about the sender, but it incorrectly states: _It is RECOMMENDED that the **receiving** endpoint explicitly reject such requests._

I think it is beneficial to retain a paragraph that concentrates on handling of the HTTP messages in transit, and therefore still think that removing this paragraph (starting from line 694) is the correct thing to do, _and also fixing the "receiving" in the sentence starting from line 708 to "sending."

Reply:
I think the confusing piece here is that the sender of the GOAWAY is the receiver of the in-flight request/push.  This means to talk about the receipt of the request, but makes it sound like it's talking about the receiver of the GOAWAY.

I'm not sure we need to eliminate any paragraphs, but clarifying the terms would probably help a lot.  This section could previously rely on "client" and "server" to clearly imply who's doing what with requests and GOAWAY frames; now it's generalized and easy to lose track of the involved parties.

Reply:
@kazuho, I've updated the text around both statements to try to make this clearer.  Can you take a look at 3f93d362?

Reply:
Thank you for all the work (including merging the changes from master). Looks good to me now.